src/System.Private.CoreLib/src/System/Text/Utf8Span.Enumeration.cs

```diff
@@ -0,0 +1,137 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Buffers;
+using System.Diagnostics;
+
+namespace System.Text
+{
+    public readonly ref partial struct Utf8Span
+    {
+        public CharEnumerable Chars => new CharEnumerable(this);
+        public RuneEnumerable Runes => new RuneEnumerable(this);
+
+        public readonly ref struct CharEnumerable
+        {
+            private readonly Utf8Span _span;
+
+            internal CharEnumerable(Utf8Span span)
+            {
+                _span = span;
+            }
+
+            public Enumerator GetEnumerator() => new Enumerator(_span);
+
+            public ref struct Enumerator
+            {
+                private uint _currentCharPair;
+                private ReadOnlySpan<byte> _remainingUtf8Bytes;
+
+                internal Enumerator(Utf8Span span)
+                {
+                    _currentCharPair = default;
+                    _remainingUtf8Bytes = span.Bytes;
+                }
+
+                public char Current => (char)_currentCharPair;
+
+                public bool MoveNext()
+                {
+                    // We don't need to worry about tearing since this enumerator is a ref struct.
+
+                    if (_currentCharPair > char.MaxValue)
+                    {
+                        // There was a surrogate pair smuggled in here from a previous operation.
+                        // Shift out the high surrogate value and return immediately.
+
+                        _currentCharPair >>= 16;
+                        return true;
+                    }
+
+                    if (_remainingUtf8Bytes.IsEmpty)
+                    {
+                        return false;
+                    }
+
+                    // TODO_UTF8STRING: Since we assume Utf8String instances are well-formed, we may instead
+                    // call an optimized version of the "decode" routine below which skips well-formedness checks.
+
+                    OperationStatus status = Rune.DecodeFromUtf8(_remainingUtf8Bytes, out Rune currentRune, out int bytesConsumed);
+                    Debug.Assert(status == OperationStatus.Done, "Somebody fed us invalid data?");
+
+                    if (currentRune.IsBmp)
+                    {
+                        // Common case - BMP scalar value.
+
+                        _currentCharPair = (uint)currentRune.Value;
+                    }
+                    else
+                    {
+                        // Uncommon case - supplementary plane (astral) scalar value.
```

Comment:
>Uncommon [](start = 27, length = 8)

what happen when having not well formed surrogate? do we throw at that time?

Reply:
`Utf8Span` instances which contain ill-formed UTF-8 data have undefined behavior. The only public ctor is one which takes a `Utf8String` (which is guaranteed to be well-formed). There's also an unsafe API which allows the caller to say "Trust me - this is well-formed data. Create the Utf8Span instance over this buffer." But it's unsafe and its contract is spelled out.